File: static/js/affiliate.js
/**
 * Affiliate (제휴업체) analysis tab - month-based view.
 * 월 기준으로 그룹핑, 각 월 내 업체별 매출/주문/객수량 표시.
 * 해당 월에 0건인 업체도 모두 노출.
 */
const Affiliate = {
    data: [],
    initialized: false,

    getDateRange() {
        const startSel = document.getElementById('affStartMonth');
        const endSel = document.getElementById('affEndMonth');
        const start = startSel.value;
        const endVal = endSel.value;
        const [ey, em] = endVal.split('-').map(Number);
        const endDate = em === 12
            ? `${ey + 1}-01-01`
            : `${ey}-${String(em + 1).padStart(2, '0')}-01`;
        return { start: start + '-01', end: endDate, startMonth: start, endMonth: endVal, label: `${start} ~ ${endVal}` };
    },

    initFilters() {
        if (this.initialized) return;
        this.initialized = true;
        const startSel = document.getElementById('affStartMonth');
        const endSel = document.getElementById('affEndMonth');
        const months = [];
        const now = new Date();
        let y = 2023, m = 1;
        while (y < now.getFullYear() || (y === now.getFullYear() && m <= now.getMonth() + 1)) {
            months.push(`${y}-${String(m).padStart(2, '0')}`);
            m++;
            if (m > 12) { m = 1; y++; }
        }
        months.forEach(mo => {
            startSel.add(new Option(mo, mo));
            endSel.add(new Option(mo, mo));
        });
        startSel.value = '2025-01';
        endSel.value = '2026-02';
    },

    async load() {
        this.initFilters();
        const { start, end, label } = this.getDateRange();
        const data = await Utils.fetchJson(`/api/affiliate/monthly?start=${start}&end=${end}`);
        this.data = data;
        this.render(label);
    },

    render(label) {
        const tbody = document.querySelector('#affiliateTable tbody');
        if (!tbody) return;

        // Collect all companies and all months
        const allMonths = new Set();
        const companies = this.data.map(comp => {
            Object.keys(comp.months || {}).forEach(m => allMonths.add(m));
            return comp;
        });
        const monthsSorted = [...allMonths].sort();

        // If no months from data, generate from filter range
        if (monthsSorted.length === 0) {
            const { startMonth, endMonth } = this.getDateRange();
            let [sy, sm] = startMonth.split('-').map(Number);
            const [ey, em] = endMonth.split('-').map(Number);
            while (sy < ey || (sy === ey && sm <= em)) {
                monthsSorted.push(`${sy}-${String(sm).padStart(2, '0')}`);
                sm++;
                if (sm > 12) { sm = 1; sy++; }
            }
        }

        let grandOrders = 0, grandRevenue = 0, grandQty = 0;
        let html = '';

        monthsSorted.forEach(month => {
            // Build rows for this month: all companies with their data (0 if missing)
            const rows = companies.map(comp => {
                const d = (comp.months || {})[month] || { orders: 0, revenue: 0, total_qty: 0 };
                return {
                    name: comp.company_name,
                    kind: comp.jaehu_kind,
                    orders: d.orders || 0,
                    revenue: d.revenue || 0,
                    qty: d.total_qty || 0,
                };
            });

            // Sort by revenue desc within month
            rows.sort((a, b) => b.revenue - a.revenue);

            const monthRevenue = rows.reduce((s, r) => s + r.revenue, 0);
            const monthOrders = rows.reduce((s, r) => s + r.orders, 0);
            const monthQty = rows.reduce((s, r) => s + r.qty, 0);
            grandOrders += monthOrders;
            grandRevenue += monthRevenue;
            grandQty += monthQty;

            // Month header row
            html += `<tr class="bg-blue-100">
                <td colspan="8" class="px-3 py-2 font-bold text-blue-800">
                    ${month} | 주문 ${Utils.fmt(monthOrders)}건 | 매출 ${Utils.fmt(monthRevenue)}원
                </td>
            </tr>`;

            // Company rows
            rows.forEach((r, idx) => {
                const share = monthRevenue ? ((r.revenue / monthRevenue) * 100).toFixed(1) : '0';
                const avgPrice = r.orders ? Math.round(r.revenue / r.orders) : 0;
                const avgQty = r.orders ? (r.qty / r.orders).toFixed(1) : '0';
                html += `<tr class="hover:bg-blue-50 border-t border-gray-100">
                    <td class="px-3 py-1 text-center">${idx + 1}</td>
                    <td class="px-3 py-1">${r.name}</td>
                    <td class="px-3 py-1 text-center">${r.kind}</td>
                    <td class="px-3 py-1 text-right">${Utils.fmt(r.orders)}</td>
                    <td class="px-3 py-1 text-right">${Utils.fmt(r.revenue)}</td>
                    <td class="px-3 py-1 text-right">${share}%</td>
                    <td class="px-3 py-1 text-right">${Utils.fmt(avgPrice)}</td>
                    <td class="px-3 py-1 text-right">${avgQty}</td>
                </tr>`;
            });
        });

        tbody.innerHTML = html;

        const gAvgPrice = grandOrders ? Math.round(grandRevenue / grandOrders) : 0;
        const gAvgQty = grandOrders ? (grandQty / grandOrders).toFixed(1) : '0';
        const periodInfo = label ? ` | ${label} (결제일)` : '';
        document.getElementById('affSummary').innerHTML =
            `전체 ${Utils.fmt(this.data.length)}개 업체 | ${monthsSorted.length}개월 | 주문 ${Utils.fmt(grandOrders)}건 | 매출 ${Utils.fmt(grandRevenue)}원 | 객단가 ${Utils.fmt(gAvgPrice)}원 | 객수량 ${gAvgQty}장${periodInfo}`;
    },

    exportUrl() {
        const { start, end } = this.getDateRange();
        return `/api/affiliate/export?start=${start}&end=${end}`;
    }
};

File: api/affiliate.py
"""바른손몰 affiliate (제휴업체) analysis endpoints."""
import io
import re
from flask import Blueprint, request, jsonify, send_file
from db.connection import query
from db.queries_affiliate import AFFILIATE_RANKING, AFFILIATE_MONTHLY

bp = Blueprint("affiliate", __name__)

_ILLEGAL_RE = re.compile(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f-\x9f]")


def _decode_bin(val):
    if val is None:
        return None
    if isinstance(val, (bytes, bytearray)):
        try:
            return val.decode("euc-kr")
        except Exception:
            return val.decode("utf-8", errors="replace")
    return str(val)


def _clean(s):
    if not s:
        return s
    return _ILLEGAL_RE.sub("", s)


@bp.route("/api/affiliate/ranking")
def affiliate_ranking():
    start = request.args.get("start", "2025-01-01")
    end = request.args.get("end", "2026-03-01")
    rows = query("bar_shop1", AFFILIATE_RANKING, (start, end))
    result = []
    for r in rows:
        name = _clean(_decode_bin(r.get("company_name_bin"))) or ""
        result.append({
            "company_seq": r["company_seq"],
            "company_name": name,
            "jaehu_kind": r["JAEHU_KIND"] or "",
            "orders": r["orders"],
            "revenue": r["revenue"] or 0,
            "total_qty": r["total_qty"] or 0,
            "avg_price": r["avg_price"] or 0,
            "avg_qty": float(r["avg_qty"]) if r["avg_qty"] else 0,
        })
    return jsonify(result)


@bp.route("/api/affiliate/monthly")
def affiliate_monthly():
    start = request.args.get("start", "2025-01-01")
    end = request.args.get("end", "2026-03-01")
    rows = query("bar_shop1", AFFILIATE_MONTHLY, (start, end))
    # Build: { company_seq: { name, months: { "2025-01": {orders, revenue} } } }
    companies = {}
    for r in rows:
        seq = r["company_seq"]
        if seq not in companies:
            name = _clean(_decode_bin(r.get("company_name_bin"))) or ""
            companies[seq] = {"company_name": name, "jaehu_kind": r["JAEHU_KIND"] or "", "months": {}, "total_revenue": 0}
        m = r["month"]
        companies[seq]["months"][m] = {
            "orders": r["orders"],
            "revenue": r["revenue"] or 0,
            "total_qty": r["total_qty"] or 0,
        }
        companies[seq]["total_revenue"] += (r["revenue"] or 0)
    # Sort by total revenue desc
    sorted_list = sorted(companies.values(), key=lambda x: x["total_revenue"], reverse=True)
    return jsonify(sorted_list)


@bp.route("/api/affiliate/export")
def affiliate_export():
    from openpyxl import Workbook
    from openpyxl.styles import Font, Alignment, Border, Side, PatternFill

    start = request.args.get("start", "2025-01-01")
    end = request.args.get("end", "2026-03-01")

    rows = query("bar_shop1", AFFILIATE_MONTHLY, (start, end))

    # Build company data
    companies = {}
    for r in rows:
        seq = r["company_seq"]
        if seq not in companies:
            name = _clean(_decode_bin(r.get("company_name_bin"))) or ""
            companies[seq] = {"name": name, "kind": r["JAEHU_KIND"] or "", "months": {},
                              "total_rev": 0, "total_ord": 0, "total_qty": 0}
        m = r["month"]
        rev = r["revenue"] or 0
        ord_ = r["orders"] or 0
        qty = r["total_qty"] or 0
        companies[seq]["months"][m] = {"orders": ord_, "revenue": rev, "qty": qty}
        companies[seq]["total_rev"] += rev
        companies[seq]["total_ord"] += ord_
        companies[seq]["total_qty"] += qty

    all_companies = list(companies.values())

    # Collect all months
    all_months = set()
    for comp in all_companies:
        all_months.update(comp["months"].keys())
    months_sorted = sorted(all_months)

    thin = Border(
        left=Side(style="thin"), right=Side(style="thin"),
        top=Side(style="thin"), bottom=Side(style="thin"),
    )
    hdr_fill = PatternFill(start_color="4472C4", end_color="4472C4", fill_type="solid")
    month_fill = PatternFill(start_color="D9E2F3", end_color="D9E2F3", fill_type="solid")
    hdr_font = Font(bold=True, size=10, color="FFFFFF")
    bold_font = Font(bold=True, size=10)
    center = Alignment(horizontal="center", vertical="center")
    right_align = Alignment(horizontal="right", vertical="center")

    wb = Workbook()
    ws = wb.active
    ws.title = "제휴업체별 월별 매출"

    headers = ["순위", "업체명", "제휴종류", "주문수", "매출액", "비중", "객단가", "객수량"]
    for c, h in enumerate(headers, 1):
        cell = ws.cell(row=1, column=c, value=h)
        cell.font = hdr_font
        cell.fill = hdr_fill
        cell.alignment = center
        cell.border = thin

    row_idx = 2

    for month in months_sorted:
        # Build rows for this month (all companies, 0 if missing)
        month_rows = []
        for comp in all_companies:
            d = comp["months"].get(month, {"orders": 0, "revenue": 0, "qty": 0})
            month_rows.append({
                "name": comp["name"], "kind": comp["kind"],
                "orders": d.get("orders", 0) or 0,
                "revenue": d.get("revenue", 0) or 0,
                "qty": d.get("qty", 0) or 0,
            })
        month_rows.sort(key=lambda x: x["revenue"], reverse=True)
        month_total_rev = sum(r["revenue"] for r in month_rows)
        month_total_ord = sum(r["orders"] for r in month_rows)

        # Month header row
        cell = ws.cell(row=row_idx, column=1, value=month)
        cell.font = bold_font
        cell.fill = month_fill
        cell.alignment = center
        for c in range(1, 9):
            ws.cell(row=row_idx, column=c).fill = month_fill
            ws.cell(row=row_idx, column=c).border = thin
        summary = f"주문 {month_total_ord:,}건 | 매출 {month_total_rev:,}원"
        cell2 = ws.cell(row=row_idx, column=2, value=summary)
        cell2.font = bold_font
        cell2.fill = month_fill
        row_idx += 1

        # Company rows
        for rank, r in enumerate(month_rows, 1):
            share = round(r["revenue"] / month_total_rev * 100, 1) if month_total_rev else 0
            avg_p = round(r["revenue"] / r["orders"]) if r["orders"] else 0
            avg_q = round(r["qty"] / r["orders"], 1) if r["orders"] else 0

            vals = [rank, r["name"], r["kind"],
                    r["orders"], r["revenue"], f"{share}%", avg_p, avg_q]
            for c, v in enumerate(vals, 1):
                cell = ws.cell(row=row_idx, column=c, value=v)
                cell.border = thin
                if c in (4, 5, 7):
                    cell.number_format = "#,##0"
                    cell.alignment = right_align
                elif c in (1, 3, 6, 8):
                    cell.alignment = center
            row_idx += 1

    # Column widths
    col_widths = {"A": 8, "B": 24, "C": 10, "D": 12, "E": 16, "F": 8, "G": 14, "H": 10}
    for letter, w in col_widths.items():
        ws.column_dimensions[letter].width = w

    ws.freeze_panes = "A2"
    ws.auto_filter.ref = f"A1:H{row_idx - 1}"

    buf = io.BytesIO()
    wb.save(buf)
    buf.seek(0)
    return send_file(buf, as_attachment=True, download_name="affiliate_monthly.xlsx",
                     mimetype="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet")
